Insert a singleton (#2062)
oasis-tcs/odata-specs: odata-protocol/11.4 Data Modification.md

Author: Wording: URL of a collection/entity vs. resource path of a collection/entity

Reviewer: I like resource path...

requests.

For requests including an [`OData-Version`](#HeaderODataVersion) header
value of `4.01`, any ETag values specified in the request body of a
[request for modifying an entity](#UpdateanEntity) MUST be `*` or match the current value
for the record being updated.

::: example

> Example ##ex_antietag: `Category` is a single-valued nullable non-containment navigation property

Reviewer: Note that, (in the given case), the etag used in an if-match or if-not-match header applies to the category currently associated with the product. if you want to make an update conditional on the product, you would instead issue the update against the product, with the updates to the category in the payload.

representation of an entity of the declared target entity type,
or one of its derived types.

If the service is unable to determine the canonical collection (as defined in [section ##ContextURL])
for the entity, it MUST fail the request.
This determination is possible, for example, if the collection's URL
- consists of an entity set
- ends with a containment navigation property

> - ends with a non-containment navigation property with a navigation property binding

Author: Is the navigation property binding really sufficient? Given
```xml
<EntitySet Name="Orders">
  <NavigationPropertyBinding Path="DeepLink" Target="ns.container/Orders/Items/Subitems" />
</EntitySet>
```
how can a `POST Orders(23)/DeepLink` determine the `x` in
```
Location: http://server/Orders(23)/Items(x)/Subitems(y)
```
for the newly-created entity? (`y` is generated by the service.)

> - ends with a non-containment navigation property with a navigation property binding

Author: Really? A navigation property binding
```xml
<EntitySet Name="Orders">
  <NavigationPropertyBinding Path="Items/CancellationItem" Target="ns.container/Orders/Items" />
</EntitySet>
```
is insufficient to determine the canonical collection for the create request `POST Orders(23)/Items(5)/CancellationItem`. This needs service-specific knowledge that the cancellation item belongs to the same order, hence the collection is `Orders(23)/Items`.

Reviewer 2: This is not a valid [binding target](https://docs.oasis-open.org/odata/odata-csdl-xml/v4.02/csd01/odata-csdl-xml-v4.02-csd01.html#BindingTarget). 

> If the target is a target path, it MUST resolve to an entity set, singleton, or direct or indirect containment navigation property of a _singleton_ in scope.

With that restriction in mind line 156 should work.

Author: The term "containment navigation property of a singleton" seems undefined. Is this
- a single-valued containment navigation property?
- a path starting with a singleton that ends in a containment navigation property?

Why are singletons treated specially here?

Reviewer 2: > The path can traverse single-valued containment navigation properties or single-valued complex properties before ending in a containment navigation property, and there MUST NOT be any non-containment navigation properties prior to the final segment.

What is missing in this definition?

Author: This definition is OK, but the additional condition
> If the target is a target path, it MUST resolve to an entity set, singleton, or direct or indirect _containment navigation property of a singleton_ in scope.

leaves unclear where the singleton must be (at the beginning or the end of the path).

Reviewer 2: Thanks! Please suggest an addition.

Reviewer: Discussed 2025-3-26 and agreed that the current wording is sufficient.

> If the service is unable to determine the canonical collection (as defined in [section ##ContextURL])
> for the entity, it MUST fail the request.
> This determination is possible, for example, if the collection's URL
> - consists of an entity set
> - ends with a containment navigation property
> - matches a navigation property path binding [#OData-CSDL#NavigationPropertyPathBinding]
> 
> or the payload contains a context URL.

Reviewer: Maybe preface this with something like "The URL specified in the POST request need not be the canonical URL for the collection; for example, it could represent a non-contained navigation property.  In such a case, ..."

And, since we have now scoped it to the non-canonical case, we don't need to talk about 1 & 2, and can just say something like:

```suggestion
The URL specified in the POST request need not be the canonical URL for the collection; for example, it could represent a non-contained navigation property. In such a case, the service can determine the canonical collection through a matching navigation property path binding [#OData-CSDL#NavigationPropertyPathBinding] or the presence of a context URL within the payload. If the service cannot determine the canonical collection, it MUST fail the request.
```

This describes the common cases, but leaves open the option for the service to have other ways to determine the canonical collection.

Author: I like that wording, will use it elsewhere as well.

The service MUST fail the request if the

> entity represented in the payload exists already in the determined collection.

Author: Or should the POST request then be treated as an update and return 200 OK without Location header?

Reviewer 2: No, POST to single entity URL is intentionally undefined.

Author: This is not about POST to a single-entity URL, but to a collection-valued non-containment navigation property, for example
```http
POST Orders(23)/Products

{"ProductID": 5}
```
if Product 5 already exists.

Reviewer 2: So far such a request always failed, and I would not change that in a minor version.

Reviewer: This seems to me like it would be contrary to REST patterns.

Reviewer: 2025-4-16: agreed to keep an error.

### ##subsubsec Create an Entity

To create an entity in a collection, the client sends a `POST` request
to a resource path addressing that collection. The `POST` body MUST contain a single valid
representation of an entity of the declared target entity type,
or one of its derived types.

The resource path need not be the canonical URL for the collection; for example,

> it could represent a non-contained navigation property.

Reviewer: ```suggestion
it could represent a non-containment navigation property.
```

> In such a case, the service can determine the canonical collection (as defined in [section ##ContextURL])
> through a navigation property path binding [#OData-CSDL#NavigationPropertyPathBinding]
> that applies to all instances addressed by the resource path
> or through the presence of a context URL within the payload.
> If the service is unable to determine the canonical collection for the entity, it MUST fail the request.

Reviewer: ```suggestion
In such a case, the service may be able to determine the canonical collection (as defined in [section ##ContextURL]), for example, through a navigation property path binding [#OData-CSDL#NavigationPropertyPathBinding]
or through the presence of a context URL within the payload.
If the service is unable to determine the canonical collection for the entity, it MUST fail the request.
```

The service MUST fail the request if the

> entity represented in the payload exists already in the determined collection.

Reviewer: ```suggestion
entity represented in the payload already exists in the determined collection.
```

references to existing entities or as content for new related entities.

> An entity may also be created as the result of an

Author: Does this cover the case of a nullable singleton? Singletons are not entities, are they?

Reviewer 2: This section is about POST to a collection URL. A singleton is _not_ a one-element collection, and OData does not define POST to a singleton URL. We had a long debate on this and decided against it.

So this sentence here does not apply to singletons. 

The referenced sentence on PUT/PATCH to a single entity URL does apply to singletons.

[Upsert](#UpsertanEntity) operation, which is then said to be "treated as an insert".
This includes the case where a nullable singleton
or single-valued navigation property is changed from null to any value

> (even the redundant change from null to null).

Reviewer 2: Why mention the change from null to null? Does this statement imply that the service cannot simply do nothing when it receives this redundant "no-op" change?

### ##subsubsec Update an Entity

To make changes to an entity, the client makes `PATCH` or `PUT`

> request (a so-called _upsert request_) to a URL that identifies the entity. The [edit URL](#ReadURLsandEditURLs)

Reviewer: I would not use upsert here; depending on whether the if-match it may just be an update.

### ##subsubsec Update an Entity

> To make changes to an entity, the client makes `PATCH` or `PUT`

Reviewer: ```suggestion
To make changes to an entity, the client makes a `PATCH` or `PUT`
```

Updating a dependent property that is tied to a key property of the
principal entity through a referential constraint updates the
relationship to point to the entity with the specified key value. If
the entity set for that entity cannot be determined or does not contain
such an entity, the update fails. The entity set can be determined, among

> other possibilities, by containment navigation properties or navigation property bindings in the
> service metadata, or by `@context` or `@id` control information in the request
> payload.

Reviewer 2: ```suggestion
other possibilities, by containment or navigation property binding,
or by control information in the request payload.
```
I'd prefer to keep specifics of the JSON Format out of the Protocol specification.

Updating a dependent property that is tied to a key property of the
principal entity through a referential constraint updates the
relationship to point to the entity with the specified key value. If

> the entity set for that entity cannot be determined or does not contain

Reviewer: should be canonical collection
```suggestion
the canonical collection for that entity cannot be determined or does not contain
```

Updating a dependent property that is tied to a key property of the
principal entity through a referential constraint updates the
relationship to point to the entity with the specified key value. If

> the entity set for that entity cannot be determined or does not contain

Author: This is already excluded because singletons have no key.

Updating a dependent property that is tied to a key property of the
principal entity through a referential constraint updates the
relationship to point to the entity with the specified key value. If
the canonical collection (as defined in [section ##ContextURL]) for that entity cannot be determined or does not contain

> such an entity, the update fails. The canonical collection can be determined, among
> other possibilities, by containment navigation properties or navigation property bindings
> or by a context URL in the request payload.

Reviewer: ```suggestion
such an entity, the update fails. The canonical collection is known for containment navigation properties, and can be determined in the presence of navigation property bindings or a context URL in the request payload, or through service specific knowledge.
```

or similar wording as above.

#### ##subsubsubsec Update Related Entities When Updating an Entity

> Requests for changing an entity with an OData-Version header with a value of `4.0` MUST

Reviewer: ```suggestion
Requests to change an entity with an OData-Version header with a value of `4.0` MUST
```
(throughout)

### ##subsubsec Upsert an Entity

An upsert occurs when the client sends an [update
request](#UpdateanEntity) to a valid URL for addressing a single entity [#OData-URL#AddressingEntities]
and the requested update newly creates the addressed entity.
In this case the service MUST handle the
request as a [create entity request](#CreateanEntity) or fail the
request altogether.

Upserts are possible if the canonical collection or canonical singleton
(as defined in [section ##ContextURL]) can be determined for the newly-created entity.
Among other possibilities, this is the case if

> 1. the URL contains the name of an entity set or singleton

Author: More precision needed? "URL contains a segment consisting of the name of an entity set or singleton"

On failure, the service MUST NOT apply any of the changes specified in
the request.

#### ##subsubsubsec Upsert an Entity

Services MAY treat an upsert request like a [create entity request](#CreateanEntity)
if it requires the addressed entity to be newly created. The request is then said to be
"treated as an insert", otherwise "treated as an update".
Services that support this capability (instead of failing upsert requests that require
entity creation) SHOULD advertise it by an annotation with the
term `Capabilities.UpdateRestrictions` (nested property `Upsertable`
with value `true`) defined in [OData-VocCap](#ODataVocCap).

If the service is unable to determine the
canonical collection or canonical singleton (as defined in [section ##ContextURL]) for the newly-created entity,
it MUST fail the upsert request.
This determination is possible, for example, if the resource path of the request
1. consists of an entity set followed by a key predicate
2. consists of a singleton (which must be nullable if the request can be treated as an insert)
3. ends with a containment navigation property,
   followed by a key predicate if this is collection-valued

>    (and which must be nullable otherwise if the request can be treated as an insert)

Reviewer: I would not try to cover both single- and collection-valued in one bullet; it makes it too confusing (especially with the part about nullability).

Reviewer: We could combine entity set and collection-value containment navigation property, and combine singleton and single-valued containment navigation property.

4. matches a navigation property path binding [#OData-CSDL#NavigationPropertyPathBinding]

> or the payload includes a context URL.

Reviewer 2: Why isn’t this the fifth numbered item in the list?

Author: Items 1-4 have the common prefix "if the resource path of the request", the fifth item is different.

Upserts are not supported against entities whose keys' values are
generated by the service. Services MUST fail an upsert request to a URL
that would identify such an entity and the entity does not yet exist.

Similarly, services MUST fail an upsert request to the URL of a [media entity](#RequestingtheMediaStreamofaMediaEntityusingvalue) that does not yet exist.
However, a `PUT` request to the _media edit URL_ of a media entity does have upsert
semantics, in that the media entity is [created](#CreateaMediaEntity)
with the specified media stream if it does not already exist, otherwise the
media stream of the existing media entity is
[updated](#UpdateaMediaEntityStream).

A key property whose value is provided in the request URL SHOULD be omitted from the request body.
If key properties are provided in the request URL and the request body with different values,
services MUST either fail the request or ignore the value in the request body.

> If the URL of an upsert request falls into case 1, 2, or 3 given above and the request

Reviewer: would rather not reference numbers here, if we can avoid it.

Reviewer: maybe specify "the resource path is the canonical url" or something of that sort (if, indeed, we agree that it is different for the canonical url versus a nav prop binding...)

- contains an [`If-Match: *`](#HeaderIfMatch) header, the service MUST fail the request
  if it would be treated as an insert

> - contains an `If-Match` header with an ETag as value, the service fails the request

Reviewer 2: Why not „the service MUST fail the request“?

Author: Because this is not a new requirement, it already follows from section 11.4.1.1.

> - contains an `If-Match` header with an ETag as value, the service fails the request
>   instead of treating it as an insert because the [ETag](#UseofETagsforAvoidingUpdateConflicts)
>   does not match

Reviewer: this leaves out the fact that the service can treat it as an update.
```suggestion
- contains an `If-Match` header with an ETag as value, the service treats the request
  as an update if the ETag matches the existing entity, otherwise fails the request
```

- contains an [`If-None-Match: *`](#HeaderIfNoneMatch) header, the service MUST fail the request
  if it would be treated as an update.

In other words, the `If-`(`None-`)`Match` header distinguishes between insert
and update in these cases.

Note outside of these cases it can happen that an upsert request does not address an existing entity with its URL
but identifies one in its payload. Such a request may update the identified entity and
link it to another entity, but it does not lead to a create entity request.

> ::: example

Reviewer: consider moving this example (or a discussion of this example) to the etag section.

Example ##ex: `Category` is a single-valued nullable non-containment navigation property
on the product entity type. Even if
```
GET http://host/service/Products(57)/Category
```
returns `404 Not Found`, the upsert request
```
PATCH http://host/service/Products(57)/Category
```

> is treated like an update not an insert if its payload identifies an existing category.

Author: This PATCH request may or may not be treated as an insert, hence the `Upsertable` capability may or may not be relevant. This means that a client cannot know it advance whether it can make this request.

On failure, the service MUST NOT apply any of the changes specified in
the request.

> #### ##subsubsubsec Upsert an Entity

Reviewer 2: Given that the Update section requires failing the request when the addressed entity does not exist, I find it confusing to make this now a subsection.

> An upsert occurs when an [update request](#UpdateanEntity) newly creates the addressed entity.

Reviewer 2: An upsert is a request that updates the target if it exists, and inserts it if it does not yet exist. This is how [HTTP defines PUT](https://www.rfc-editor.org/rfc/rfc9110.html#name-put), and similar to the [MODIFY statement in ABAP](https://help.sap.com/doc/abapdocu_751_index_htm/7.51/de-DE/abapmodify_dbtab.htm).

The client doesn't know or care whether the target already exists, and the server does the needful.

If the client cares, it adds `If-Match: *` or `If-None-Match: *` to the request.

"Upsert" is the *client intention*, and "update or insert" is the *server reaction*.

Author: Let's decide the terminology in the TC.
What|Heiko|Ralf
---|---|---
client intention|update|upsert
if entity exists|no upsert|update
if entity is created|upsert|insert

Reviewer 2: https://en.wiktionary.org/wiki/upsert

https://wiki.postgresql.org/wiki/UPSERT

https://www.cockroachlabs.com/blog/sql-upsert/

On failure, the service MUST NOT apply any of the changes specified in
the request.

#### ##subsubsubsec Upsert an Entity

> Services MAY treat a request for changing an entity like a [create entity request](#CreateanEntity)

Reviewer: ```suggestion
Services MAY treat a request to change an entity as a [create entity request](#CreateanEntity)
```

if it requires the addressed entity to be newly created. The request is then said to be
"treated as an insert", otherwise "treated as an update".
Services that support this "upsert" capability (instead of failing such requests that require
entity creation) SHOULD advertise it by an annotation with the
term `Capabilities.UpdateRestrictions` (nested property `Upsertable`
with value `true`) defined in [OData-VocCap](#ODataVocCap).

If resource path is not a canonical URL and the request is treated as an insert,
the service can determine the canonical collection or canonical singleton (as defined in [section ##ContextURL]) for the newly-created entity
through a navigation property path binding [#OData-CSDL#NavigationPropertyPathBinding]
that applies to the instance addressed by the resource path
or through the presence of a context URL within the payload;
if the service is unable to determine the canonical collection, it MUST fail the request.

Upserts are not supported against entities whose keys' values are
generated by the service. Services MUST fail a request to a URL
that would identify such an entity and the entity does not yet exist.

> Similarly, services MUST fail a `PATCH` request to the URL of a [media entity](#RequestingtheMediaStreamofaMediaEntityusingvalue) that does not yet exist.

Reviewer: ```suggestion
Similarly, services MUST fail an update request to the URL of a [media entity](#RequestingtheMediaStreamofaMediaEntityusingvalue) that does not yet exist.
```

media stream of the existing media entity is
[updated](#UpdateaMediaEntityStream).

A key property whose value is provided in the request URL SHOULD be omitted from the request body.
If key properties are provided in the request URL and the request body with different values,
services MUST either fail the request or ignore the value in the request body.

If the resource path is a canonical URL and the request
- contains an [`If-Match: *`](#HeaderIfMatch) header, the service MUST fail the request
  if it would be treated as an insert
- contains an `If-Match` header with an ETag as value, the service treats the request as
  an update if the [ETag](#UseofETagsforAvoidingUpdateConflicts) matches the addressed entity
  or fails the request, but never treats it as an insert
- contains an [`If-None-Match: *`](#HeaderIfNoneMatch) header, the service MUST fail the request
  if it would be treated as an update.

In other words, the `If-`(`None-`)`Match` header distinguishes between insert
and update in this case. Otherwise the resource path ends with a
non-containment navigation property and an `If-`(`None-`)`Match` header applies to

> the current related entity addressed by it (see [example ##antietag]).

Author: In the other case, the client may not care whether the operation is an update or upsert, and hence may not want to evaluate both `Updatable` and `Upsertable` in the capabilities annotations.

with value `true`) defined in [OData-VocCap](#ODataVocCap).

A key property whose value is provided in the request URL SHOULD be omitted from the request body.
If key properties are provided in the request URL and the request body with different values,
services MUST either fail the request or ignore the value in the request body.

If the URL of an update request falls into case 1, 2, or 3 given above and the request
- contains an [`If-Match: *`](#HeaderIfMatch) header, the service MUST fail the request
  if it would newly create the addressed entity in the absence of the header
- contains an `If-Match` header with an etag as value, the service fails the request
  instead of newly creating the addressed entity because the [etag](#UseofETagsforAvoidingUpdateConflicts)
  does not match
- contains an [`If-None-Match: *`](#HeaderIfNoneMatch) header, the service MUST fail the request
  if it would update an existing entity in the absence of the header.

In other words, the `If-`(`None-`)`Match` header distinguishes between update and upsert
in these cases.

If the URL of an update request does not fall into case 1, 2, or 3 given above, it ends

> with a non-containment navigation property, optionally followed by a key predicate.

Author: In this case, the client may not care whether the operation is an update or upsert, and hence may not want to evaluate both `Updatable` and `Upsertable` in the capabilities annotations.
